### ISSUE TITLE
[React-Typescript] Return type is forced as any without explicit declaration

### ISSUE THREAD

**Issue by FredyC** [teambit/bit.envs#53]
For example, this exported React hook.

https://bit.dev/speedlo/hooks/use-router-navigate/~code

As you can see there is no explicit return type so TS would infer it.

However, when installed, the `useRouterNavigate.d.ts` looks like following. Notice the `any` return type. 

```ts
interface IOptions {
    to?: string;
    replace?: boolean;
}
export declare function useRouterNavigate({ to, replace }: IOptions): any;
export {};
```

The same problem exists in compiled files when exporting.

This is obviously wrong, we don't want to be typing explicitly. I've been looking over `tsconfig.json` and I am not sure which option could be causing something like that. Never seen such behavior.

Tried version 3.0.18 (no luck) and 3.0.19 is broken (#54)

**Comment by FredyC** [comment]
For the reference, when I build that code with the following settings, it does add explicit return type too, but it's actually correct one.

```ts
export declare function useRouterNavigate({ to, replace }: IOptions): (opts: IOptions) => import("history").Location<any>;
```

```json
"compilerOptions": {
        "allowJs": false,
        "noEmit": true,
        "jsx": "react",
        "target": "es5",
        "lib": [
            "esnext",
            "dom"
        ],
        "moduleResolution": "node",
        "newLine": "lf",
        "forceConsistentCasingInFileNames": false,
        "allowSyntheticDefaultImports": true,
        "allowUnreachableCode": false,
        "noUnusedLocals": true,
        "esModuleInterop": true,
        "skipLibCheck": true,
        "strictFunctionTypes": true,
        "strictNullChecks": true,
        "strictPropertyInitialization": true,
        "resolveJsonModule": true,
        "isolatedModules": true,
        "strict": false,
        "downlevelIteration": true,
        "noImplicitReturns": true,
        "noImplicitAny": false,
        "noImplicitThis": true,
        "module": "esnext"
    },
```

**Comment by qballer** [comment]
Right now we aren't reading the configuration in your project. The config is hard coded in the compiler it's self. You can see it here: https://bit.dev/bit/envs/compilers/typescript/~code#src/tsconfig.ts

Another thing - react-typescript is just a name wrapper for typescript. In case you need to fork to change your configuration. Use typescript directly, it would be easier. 

The typescript compiler is developed in this repo: https://github.com/teambit/envs. All environments made by bit will be moved there eventually.

**Comment by qballer** [comment]
in regards to types dependencies. Can you attach the result of `bit show [component-name]`

**Comment by FredyC** [comment]
Well, we misunderstood heavily each other 😆 

I don't need to use my own `tsconfig`, it was just a wild guess it might be caused by some setting in it so I've shown mine which I am using to essentially bundle a whole collection into a distinct NPM module.

I think that problem is somewhere along with finding `@types` packages, specifically `@types/history` in this case which should have been used there but TS probably did not find it for some reason so it used `any` in the output. I do have that package in `node_modules` of root folder and as I said, when I build it directly with TSC, it does output correct declaration file.

```
╔═══════════════════╤════════════════════════════════════════════════════╗
║ Id                │ use-router-navigate@0.2.0                          ║
╟───────────────────┼────────────────────────────────────────────────────╢
║ Compiler          │ bit.envs/compilers/react-typescript@3.0.18         ║
╟───────────────────┼────────────────────────────────────────────────────╢
║ Language          │ javascript                                         ║
╟───────────────────┼────────────────────────────────────────────────────╢
║ Main File         │ src/useRouterNavigate.ts                           ║
╟───────────────────┼────────────────────────────────────────────────────╢
║ Dependencies      │ speedlo.hooks/use-router-context@0.2.0             ║
╟───────────────────┼────────────────────────────────────────────────────╢
║ Dev Dependencies  │                                                    ║
╟───────────────────┼────────────────────────────────────────────────────╢
║ Peer Dependencies │ react@>=16.8                                       ║
╟───────────────────┼────────────────────────────────────────────────────╢
║ Files             │ src/useRouterNavigate.ts                           ║
╚═══════════════════╧════════════════════════════════════════════════════╝
```

**Comment by qballer** [comment]
Do you import the history package in your component? (useRouterNavigate.ts )

**Comment by FredyC** [comment]
I don't, it's indirectly imported by `react-router` which is imported by `useRouterContext` hook. 

```ts
import React from 'react'
import * as Router from 'react-router'

export function useRouterContext<Params extends Record<string, string> = {}>(): Router.RouteComponentProps<Params> {
  return React.useContext(Router.__RouterContext)
}
```

There is even an explicit return type in this case.

I am not Gitter if you want a more active conversation.

**Comment by qballer** [comment]
Can you provide a test repo? I can't build with your config. Please also commit the .bitmap file.

**Comment by qballer** [comment]
I've compiler this code base with typescript on my local machine and it seems to produce the same faulty typings. I've added
```js
   "@types/jest": "^24.0.18",
    "@types/react": "^16.9.2",
```
I've also used this configuration: 
```
{
    "compilerOptions": {
        "allowJs": false,
        "noEmit": false,
        "jsx": "react",
        "target": "es5",
        "lib": [
            "esnext",
            "dom"
        ],
        "outDir": "./dist",
        "moduleResolution": "node",
        "newLine": "lf",
        "forceConsistentCasingInFileNames": false,
        "allowSyntheticDefaultImports": true,
        "allowUnreachableCode": false,
        "noUnusedLocals": true,
        "esModuleInterop": true,
        "skipLibCheck": true,
        "strictFunctionTypes": true,
        "strictNullChecks": true,
        "strictPropertyInitialization": true,
        "resolveJsonModule": true,
        "isolatedModules": false,
        "strict": false,
        "downlevelIteration": true,
        "noImplicitReturns": true,
        "noImplicitAny": false,
        "noImplicitThis": true,
        "module": "esnext"
    },
}
```
Seems like it's the same result as our env. What am I missing?

**Comment by qballer** [comment]
Closing for now, if you have a repo which typescript latest can reconstruct that type, I will debug the issue.

**Comment by FredyC** [comment]
Why do you need the latest Typescript? It's rather problematic to upgrade it. As I said twice above already it works correctly on older version (3.2) when building directly through TSC. There is no reason to require the newest TS since the obvious culprit lies somewhere in a Bit build process.

**Comment by qballer** [comment]
it gives me the same typings for 3.2.4. can you add a tar file of your project?

**Comment by qballer** [comment]
Hey @FredyC, I just wanted to jump in and let you know I hadn't had the time to address this in the past week. I would do it next week.

**Comment by qballer** [comment]
It looks like the versions your `.bitmap` contains were never exported even though tagged. All those components seem to be a version "down" in the bit.dev website. So if `0.2.1` is missing, you are going to have `0.2.0` . I think a bit exported slipped by you.

**Comment by qballer** [comment]
OK, after I played around with versions I can see whats the issue. It looks like bit doesn't install devDependencies of component dependencies while isolating the component. Therefor typescript compiler can't infer the type. Still the best type I get is the following: 
```typescript
function useRouterNavigate({ to, replace }: IOptions): (opts: IOptions) => Location<any>
``` 
I will see how we can fix that:

[repo.tar.gz](https://github.com/teambit/bit.envs/files/3639660/repo.tar.gz)

If you download this file and run the following command, you can see that the isolation doesn't have these types install in the capsule.
```env DEBUG=true bit build use-router-navigate -c```

cc: @GiladShoham @davidfirst

**Comment by FredyC** [comment]
Thank a lot for investigating this. I thought it was something along those lines. Yea, it makes total sense that bit should install `devDependencies` as well.

Just in case you wonder why I don't add those types packages to `dependencies`, it's because same `package.json` would be used for a full package in NPM and from my experience, it can cause weird issues if multiple versions of types packages are installed.

Regarding the `.bitmap`, you are right. Looking now it was indeed taken shortly after I've tagged, but not exported. After untagging and copying that `.bitmap` the `bit import` works just fine.

**Comment by qballer** [comment]
This isn't a simple answer because when I compile `use-router-context` it does have the proper dependencies and it build properly. I think `@types/react-router` should be a devDependency of `use-router-navigate`. You can achieve it with override.

**Comment by FredyC** [comment]
Hm, what about Bit installing `@types` namespaced packages automatically (for build purposes) when seen in package.json? Ideally, if that could be declared on compiler level.

Thing is that these dependencies are usually not explicit as they are used indirectly by packages. It cannot be properly analyzed from the code itself. I can imagine in some more complex scenarios it might be quite annoying to figure out correct overrides.

Besides, it's not exactly apparent and the user has to know what to look for and build is passing. User will notice that way later when a component is installed and used. It's fairly sneaky error yielding `any`.

**Comment by qballer** [comment]
I agree, this is very sneaky. A solution which install all devDependencies may lead to "bringing" many dependencies with a component. The kind you might not need. This can; also be a bug in the typescript compiler (the part made by Microsoft). I would expect that such a composition would work, but the type inference doesn't compose well. 

I will consult with other people on the team. I'm glad though you can move forward for now with overrides. I will consult with other team members and get back to you.

**Comment by FredyC** [comment]
Perhaps I am misunderstanding something, but when Bit does the build, why it cannot just have all dependencies installed? It's not packaged with the component, just used to make a build, right?

Also, I just realized, that when a component will be installed by the consumer, he won't know to install any `@types` packages because those won't be in `package.json`. Ideally, those should be in `peerDependencies`, but since the component code doesn't directly depend on them, Bit won't include them in there, right?

**Comment by GiladShoham** [comment]
@FredyC 
since this pr https://github.com/teambit/bit/pull/1904
Bit will add the @types/X package when it appears in the package.json and X is required by a ts component.
So usually the consumer won't need to really handle it himself.
I think the problem starts when it's a dependency of the component that requires X

**Comment by qballer** [comment]
We have a rule that if `X` is a dependency of a component in the author environment also has `@types/X` then it becomes a devDependency of the component. In our case X (a.k.a react-router)  is a dependency of another dependency (context). While we compile another (navigate). This is what drives the complexity here - context has the right `@types` dependencies but there is some type information we loss. 

I would be glad to solve it in the compiler level, but this is trivial. I can "scan" the author environment, but what happens when we import.

**Comment by FredyC** [comment]
Ok, I have tried the latest CLI 14.4.1 with typescript compiler 3.0.35, but it still has trouble discovering that `@types/react-router` dependency. I had to use overrides. 

And another use case where the component is using `/// <reference types="customTypesPackage" />` for some ambient types. Obviously, that wasn't recognized either and failed the build. So that was another override.

I am not sure if this behavior about `@types` is documented somewhere, should be for sure.

I can imagine typescript components are not that common yet, especially ones that depend on each other. It might become a more relevant issue sometime in the future, but I guess it's not that important now.
